onnxruntime/core/framework/session_state.cc

```diff
@@ -156,20 +186,138 @@ static int64_t CalculateMemoryPatternsKey(const std::vector<std::reference_wrapp
   return key;
 }
 
-const MemoryPatternGroup* SessionState::GetMemoryPatternGroup(
-    const std::vector<std::reference_wrapper<const TensorShape>>& input_shapes) const {
+namespace {
+Status ResolveDimParams(const GraphViewer& graph, const std::map<std::string, TensorShape>& feeds, std::unordered_map<std::string, int64_t>& out) {
+  for (size_t i = 0; i < graph.GetInputs().size(); ++i) {
+    auto* input = graph.GetInputs()[i];
+    auto* shape = input->Shape();
+    auto it = feeds.find(input->Name());
+    if (it == feeds.end())
+      return Status(ONNXRUNTIME, FAIL, "Graph input " + input->Name() + " is not found in the feed list, unable to resolve the value for dynamic shape.");
+    if (!shape || shape->dim_size() != static_cast<int>(it->second.NumDimensions()))
+      return Status(ONNXRUNTIME, FAIL, "Graph input " + input->Name() +
+                                           "'s shape is not present or its shape doesn't match feed's shape."
+                                           "Unable to resolve the value for dynamic shape");
+    for (int k = 0; k < shape->dim_size(); ++k) {
+      if (shape->dim()[k].has_dim_param()) {
+        out.insert({shape->dim()[k].dim_param(), it->second.GetDims()[k]});
+      }
+    }
+  }
+  return Status::OK();
+}
+}  // namespace
+
+Status SessionState::GeneratePatternGroupCache(const std::vector<std::reference_wrapper<const TensorShape>>& input_shape,
+                                               const std::vector<int>& feed_mlvalue_idxs,
+                                               MemoryPatternGroup* output) const {
+  std::map<std::string, TensorShape> feeds;
+  for (size_t i = 0; i < feed_mlvalue_idxs.size(); ++i) {
```

Comment:
Same comment applies in many places like ResolveDimParams. If the end condition value isn't changing in the loop, set it up once at the start so you're not relying on the compiler to avoid the cost of a function call each iteration.  

---
In reply to: [396861676](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r396861676) [](ancestors = 396861676)

Reply:
good point, for now, we'll limit the change scope to current file. And if we notice the same when addressing comments in other files, we'll apply the same

---
In reply to: [397019190](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397019190) [](ancestors = 397019190,396861676)